pkg/sentry/fsimpl/devpts/line_discipline.go

```diff
@@ -120,6 +124,10 @@ func newLineDiscipline(termios linux.KernelTermios) *lineDiscipline {
 	return &ld
 }
 
+func (l *lineDiscipline) SetForegroundProcessGroup(pg *kernel.ProcessGroup) {
+	l.fgProcessGroup = pg
```

Comment:
The task's foregroundProcessGroup can also be changed by an ioctl to the pty, so keeping track of it in the lineDiscipline can get out-of-sync:

https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/fsimpl/devpts/master.go;l=184;drc=0927892f49b9415bddd5ea4ead2ad53654636521

It might be better to give lineDiscipline a reference to its associated Terminal, and then call Terminal.get/setForegroundProcessGroup, instead of duplicating the state.

https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/fsimpl/devpts/terminal.go;l=96;drc=0927892f49b9415bddd5ea4ead2ad53654636521

Reply:
Yes, but this PR currently already makes it so that the `SetForegroundProcessGroup`  function in `lineDiscipline`  is called by the `setForegroundProcessGroup` in the `terminal.go`.

Is there still a way that it can get out-of-sync that I'm missing?

Calling the `Terminal.get/setForegroundProcessGroup` like you suggests means that I also need to add a reference to the `ctx` in the `lineDiscipline`, which the `Terminal` has no notion of now ?

Or am I completely misunderstanding what you mean ?

Reply:
You are right, I missed the call to `setForegroundProcessGroup` in `terminal.go`.

But I'm still a bit uncomfortable with duplicating the foreground process group information in two places, since that seems unnecessary.

I think a better fix will be to plumb the context into `transformer.transform()`.

`transform()` is only called by `pushWaitBufLocked()`, and all callers of `pushWaitBufLocked()` have a context already (note that `kernel.Task` implements context):
https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/fsimpl/devpts/queue.go;l=209;drc=c020e2de836ae2a89ab01c896238da48e69524ec

 Then, if we also store the `Terminal` inside `lineDiscipline`, we can get the process group with something like `Terminal.getForegroundProcessGroup` and send the signal to that.  (Except we need a slightly different version of `getForegroundProcessGroup` that returns the process group, instead of copying out the id.) 

Let me know if this plan sounds OK.  It's more refactoring, but I think will be better in the long run.

Reply:
@nlacasse I'd looked into your suggestion and implemented it, but got a few issues.
- `writeBytes` didn't have a context (and was also calling `pushWaitBufLocked`)
- passing the context didn't really matter in the end because we didn't have the foreground processGroup available, no added value of adding it. (when implemented like this the ctrl-c SIGINT'ed the first spawned process instead the foreground one.

Instead I've added a commit more in line with your first comment.
- added a reference in `lineDiscipline` to `Terminal` now and kept the `fgProcessGroup` state in `Terminal`
- now let lineDiscipline use the fgProcessGroup to send the signals

WDYT?